src/main/scala/uk/gov/nationalarchives/omega/jms/JmsRRClient.scala

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (c) 2023 The National Archives
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.gov.nationalarchives.omega.jms
+
+import cats.data.NonEmptyList
+import cats.effect.implicits.genSpawnOps
+import cats.effect.kernel.Sync
+import cats.effect.{Async, Resource}
+import jms4s.JmsAcknowledgerConsumer.AckAction
+import jms4s.activemq.activeMQ
+import jms4s.activemq.activeMQ.{Password, Username}
+import jms4s.config.QueueName
+import jms4s.jms.{JmsMessage, MessageFactory}
+import jms4s.sqs.simpleQueueService
+import jms4s.sqs.simpleQueueService.{Credentials, DirectAddress, HTTP}
+import jms4s.{JmsClient, JmsProducer}
+import org.typelevel.log4cats.Logger
+import uk.gov.nationalarchives.omega.jms.JmsRRClient.ReplyMessageHandler
+
+import java.util.UUID
+import java.util.concurrent.ConcurrentHashMap
+import scala.concurrent.duration.DurationInt
+import scala.util.{Failure, Success}
+
+case class RequestMessage(body: String, sid: String)
+case class ReplyMessage(body: String)
+
+/**
+ * A JMS Request-Reply client.
+ * Suitable for being packaged as a library for abstracting
+ * away the complexities of JMS and jms4s.
+ *
+ * @author <a href="mailto:adam@evolvedbinary.com">Adam Retter</a>
+ */
+class JmsRRClient[F[_]: Async: Logger](requestMap: ConcurrentHashMap[String, ReplyMessageHandler[F]])(consumer: Unit, producer: JmsProducer[F]) {
+
+  def request(requestQueue: String, jmsRequest: RequestMessage, replyMessageHandler: ReplyMessageHandler[F])(implicit F: Async[F]) : F[Unit] = {
```

Comment:
Do you think it would make sense for the `replyMessageHandler` to be passed into the `JmsRRClient` constructor, rather than this method? Is there a case where we would want to change the response handler per request send?

Reply:
I suppose we need the queue.